ec2/ebs.py

```diff
@@ -0,0 +1,371 @@
+    ctx.logger.info(
+        'Attached EBS volume {0} with instance {1}.'
+        .format(volume_id, instance_id))
+
+
+@operation
+def detach(args, **_):
+    """ Detaches an EBS Volume created by Cloudify from an EC2 Instance
+    that was also created by Cloudify.
+    """
+
+    volume_id = \
+        utils.get_external_resource_id_or_raise(
+            'detach volume', ctx.source.instance)
+    instance_id = \
+        utils.get_external_resource_id_or_raise(
+            'detach volume', ctx.target.instance)
+
+    if _detach_external_volume_or_instance():
+        return
+
+    ctx.logger.debug('Detaching EBS volume {0}'.format(volume_id))
+
+    volume_object = _get_volumes_from_id(volume_id)
+    if not volume_object:
+        raise NonRecoverableError(
+            'EBS volume {0} not found in account.'.format(volume_id))
+
+    try:
+        detached = volume_object.detach(**args)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('{0}'.format(str(e)))
+
+    if not detached:
+        raise NonRecoverableError(
+            'Failed to detach volume {0} from instance {1}'
+            .format(volume_id, instance_id))
+
+    utils.unassign_runtime_property_from_resource(
+        'instance_id', ctx.source.instance)
+    ctx.logger.info(
+        'Detached volume {0} from instance {1}.'
+        .format(volume_id, instance_id))
+
+
+@operation
+def create_snapshot(args, **_):
+    """ Create a snapshot of an EBS Volume
+    """
+
+    volume_id = \
+        utils.get_external_resource_id_or_raise(
+            'create snapshot', ctx.instance)
+
+    ctx.logger.info(
+        'Trying to create a snapshot of EBS volume {0}.'
+        .format(volume_id))
+
+    volume_object = _get_volumes_from_id(volume_id)
+
+    if not args:
+        snapshot_desc = \
+            unicode(datetime.datetime.now()) + \
+            ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID]
+        args = dict(description=snapshot_desc)
+
+    try:
+        new_snapshot = volume_object.create_snapshot(**args)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('{0}'.format(str(e)))
+
+    ctx.logger.info(
+        'Created snapshot of EBS volume {0}.'.format(volume_id))
+
+    if snapshot_attribute not in ctx.instance.runtime_properties:
+        ctx.instance.runtime_properties[snapshot_attribute] = list()
+
+    ctx.instance.runtime_properties[snapshot_attribute].append(new_snapshot.id)
+
+
+def _delete_volume(volume_id):
+    """Tries to delete an EBS volume
+    """
+
+    volume_to_delete = _get_volumes_from_id(volume_id)
+
+    if not volume_to_delete:
+        raise NonRecoverableError(
+            'Unable to delete EBS volume {0}, because the volume id '
+            'does not exist in the account'.format(volume_id))
+
+    try:
+        volume_to_delete.delete()
```

Comment:
`
output = volume_to_delete.delete()
if not output:
    raise NonRecoverableError()
`